File: 130.high_level_syntax/src/main/java/HighLevelSyntax.java
package javax;

import java.util.Random;
import java.util.List;
import java.util.LinkedList;

interface Callback {
  void func(String str);
}

class MyCallback implements Callback {
  @Override
  public void func(final String str) {
    System.out.print(str);
  }
}

public class HighLevelSyntax {

  private static void functionAsParameter(final Callback callback, final String str) {
    callback.func(str);
  }

  private static Callback functionAsReturnType(final Callback callback) {
    return callback;
  }

  public static void main(String[] args) {
    final String hello = "Hello";
    final String space = " ";
    final String world = "World!";

    final List<String> words = new LinkedList<String>();

    words.add(hello);
    words.add(space);
    words.add(world);

    Callback myCallback = new MyCallback();
 
    functionAsParameter(myCallback, words.get(0));
    functionAsParameter(myCallback, words.get(1));
    functionAsReturnType(myCallback).func(words.get(2));

    System.out.println();
  }
}

File: 130.high_level_syntax/src/main/scala/scalax/HighLevelSyntax.scala
package scalax

object HighLevelSyntax extends App {
  val hello = "Hello"
  val space =  " "
  val world = "World!"

  val words = List(hello, space , world)

  val printFunc: (String) => Unit = print

  def word(n: Int) : String = {
    words(n)
  }

  // PART 1
  // Add a parameter 'func' of type function, which takes a 'String' as a parameter and returns a 'Unit'
  // Add a second parameter 'str' of type 'String'
  // Hint: See printFunc above
  def functionAsParameter() {
    // Remove the comments below i.e. the code should be executed
    // func(str)
  }

  // PART 2
  // Add a parameter 'func' of type function, which takes a 'String' as a parameter and returns a 'Unit'
  // Add a return type of type function, which takes a 'String' as a parameter and returns a 'Unit'
  // Hint: See printFunc above
	def functionAsReturnType() {
    // Remove the comments below i.e. the code should be executed
    // func
  }


  // PART 3
  // Remove the comments below i.e. the code should be executed
  // functionAsParameter(printFunc, word(0))
  // functionAsParameter(printFunc, word(1))
  // functionAsReturnType(printFunc)(word(2))

  println()
}
